### DISCUSSION TITLE
How are HTTP/2 connections managed?

### DISCUSSION THREAD

**Discussion by JustinBeckwith** [sindresorhus/got#1316] [Q&A] [answered]
Greetings!  Lovely library - I'm a big fan :) I was reading the docs, and saw that HTTP/2 is supported via a simple flag - rad!  I'm curious: 

- How are you managing the underlying `ClientHttp2Session` objects?  
- If I make two HTTP requests in a row, is it going to re-use the same underlying connection?  
- If you are re-using connections, how do you know when to close the session?

The biggest advantage I see to using an HTTP/2 client over HTTP/1.1 is taking advantage of the multiplexing over an existing TCP/IP connection.  I'd love to understand how this works.  

Full disclosure - I'm trying to figure this out in [gaxios](https://github.com/googleapis/gaxios) as well.

**Comment by szmarczak** [comment, marked as answer]
Sorry for late reply, I haven't been able to do much coding in the past month. I'm more than happy to answer your questions.

> How are you managing the underlying ClientHttp2Session objects?

I could talk about this for ages: [http2-wrapper/source/agent.js](https://github.com/szmarczak/http2-wrapper/blob/master/source/agent.js)
If you got any specific questions, please ask right away.

> If I make two HTTP requests in a row, is it going to re-use the same underlying connection?

Yes. There is [a queue](https://github.com/szmarczak/http2-wrapper/blob/ab14d1c9393c783e4cb511e9f6b3c9c7dde641b7/source/agent.js#L264-L273).

> If you are re-using connections, how do you know when to close the session?

Good question. [If a session's origin set is a subset of another session's origin set, it's closed.](https://github.com/szmarczak/http2-wrapper/blob/ab14d1c9393c783e4cb511e9f6b3c9c7dde641b7/source/agent.js#L83-L124) Another example is [when `maxFreeSessions` option is set to `0`](https://github.com/szmarczak/http2-wrapper/blob/ab14d1c9393c783e4cb511e9f6b3c9c7dde641b7/source/agent.js#L455-L461). A session can be created when there were no free sessions, and when it connected, another session can steal the listeners from the queue. [In that case, the session is closed.](https://github.com/szmarczak/http2-wrapper/blob/ab14d1c9393c783e4cb511e9f6b3c9c7dde641b7/source/agent.js#L462-L465) If a session has cleared all its requests and all the seats for free sessions are taken, [it's closed too](https://github.com/szmarczak/http2-wrapper/blob/ab14d1c9393c783e4cb511e9f6b3c9c7dde641b7/source/agent.js#L522-L526). Also there is [a timeout applied](https://github.com/szmarczak/http2-wrapper/blob/ab14d1c9393c783e4cb511e9f6b3c9c7dde641b7/source/agent.js#L357-L360).

> The biggest advantage I see to using an HTTP/2 client over HTTP/1.1 is taking advantage of the multiplexing over an existing TCP/IP connection. I'd love to understand how this works.

Got is using [`http2-wrapper`](https://github.com/szmarczak/http2-wrapper) which provides an API which is almost the same as the HTTP/1 Node.js API. There's also [an API for automatic ALPN negotiation](https://github.com/szmarczak/http2-wrapper#http2autourl-options-callback), so it knows when to use HTTP/1 or HTTP/2.

> Full disclosure - I'm trying to figure this out in gaxios as well.

Amazing! Please note that it doesn't support h2c (unsecure HTTP2).

Note: [`node-fetch` considered the use of `http2-wrapper`.](https://github.com/node-fetch/node-fetch/pull/796#issuecomment-632054711) If you share their view, you may want to use [`fetch-h2`](https://github.com/grantila/fetch-h2). Beware that they don't support multiple origins (cuz few websites do). Also they don't support multiple sessions either, so if you connect to a site that says "hey you cannot do more than 100 pending requests per session" (`maxConcurrentStreams` SETTINGS property) and exceed that limit, you will end up waiting till the first request(s) are finished.

OTOH, `http2-wrapper` session management is [heavily tested](https://github.com/szmarczak/http2-wrapper/blob/master/test/agent.js). The only things that are missing for now are TypeScript support and detailed proxy example.
